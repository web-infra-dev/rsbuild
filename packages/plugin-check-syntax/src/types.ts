import type { ecmaVersion as EcmaVersion } from 'acorn';

export type { EcmaVersion };

export type CheckSyntaxOptions = {
  /**
   * The target browser range of the project.
   * Its value is a standard browserslist array.
   */
  targets?: string[];
  /**
   * The minimum ECMAScript syntax version that can be used in the build artifact.
   * The priority of `ecmaVersion` is higher than `targets`.
   */
  exclude?: RegExp | RegExp[];
  /**
   * Used to exclude a portion of files during detection.
   * You can pass in one or more regular expressions to match the paths of source files.
   */
  ecmaVersion?: EcmaVersion;
};

export type CheckSyntaxExclude = NonNullable<CheckSyntaxOptions['exclude']>;

export interface Location {
  line: number;
  column: number;
}

export interface File {
  path: string;
  line: number;
  column: number;
}

interface SyntaxErrorOptions {
  source: File & { code: string };
  output?: File;
}

export class ECMASyntaxError extends Error {
  source: File & { code: string };

  output: File | undefined;

  constructor(message: string, options: SyntaxErrorOptions) {
    super(message);
    this.source = options.source;
    this.output = options.output;
  }
}

export type AcornParseError = {
  message: string;
  pos: number;
  loc: Location;
};
