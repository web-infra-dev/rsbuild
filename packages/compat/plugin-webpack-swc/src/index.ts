export { default as SwcLoader } from './loader';
export { pluginSwc } from './plugin';
export * from './binding';
export * from './types';
