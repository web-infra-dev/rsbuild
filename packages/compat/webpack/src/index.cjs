/**
 * CommonJS wrapper
 */
module.exports.webpackProvider = (...args) =>
  import('./index.js').then((i) => i.webpackProvider(...args));
