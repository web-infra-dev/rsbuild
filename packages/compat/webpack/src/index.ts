export { webpackProvider } from './provider';
export type { WebpackProvider } from './provider';
export { webpackBuild } from './core/build';
export { createDefaultConfig } from './config/defaults';
export type {
  LessLoaderOptions,
  SassLoaderOptions,
  PostCSSLoaderOptions,
  CSSLoaderOptions,
  StyleLoaderOptions,
} from '@rsbuild/shared';
export type {
  RsbuildPlugin,
  RsbuildPluginAPI,

  // Config Types
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  OutputConfig,
  SourceConfig,
  RsbuildConfig,
  SecurityConfig,
  NormalizedConfig,
  PerformanceConfig,
  ExperimentsConfig,

  // Hook Callback Types
  ModifyWebpackChainFn,
  ModifyWebpackConfigFn,
  ModifyWebpackChainUtils,
  ModifyWebpackConfigUtils,

  // Third Party Types
  webpack,
  WebpackChain,
  WebpackConfig,
  CSSExtractOptions,
  HTMLPluginOptions,
} from './types';
