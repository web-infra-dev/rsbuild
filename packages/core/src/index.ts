/**
 * The methods and types exported from this file are considered as
 * the public API of @rsbuild/core.
 */
import { rspack } from '@rspack/core';
import type * as Rspack from '@rspack/core';
import * as __internalHelper from './internal';

// Core methods
export { loadEnv } from './loadEnv';
export { createRsbuild } from './createRsbuild';
export { loadConfig, defineConfig } from './config';

// Rsbuild version
export const version = RSBUILD_VERSION;

// Rspack instance
export { rspack };
export type { Rspack };

// Helpers
export { logger } from '@rsbuild/shared';
export { mergeRsbuildConfig } from './mergeConfig';

// Constants
export { PLUGIN_SWC_NAME, PLUGIN_CSS_NAME } from './constants';

// Types
export type {
  // Config Types
  RsbuildConfig,
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  SourceConfig,
  ServerConfig,
  OutputConfig,
  SecurityConfig,
  PerformanceConfig,
  ModuleFederationConfig,
  // Normalized Config Types
  NormalizedConfig,
  NormalizedDevConfig,
  NormalizedHtmlConfig,
  NormalizedToolsConfig,
  NormalizedSourceConfig,
  NormalizedServerConfig,
  NormalizedOutputConfig,
  NormalizedSecurityConfig,
  NormalizedPerformanceConfig,
  NormalizedModuleFederationConfig,
  // Plugin Types
  RsbuildPlugin,
  RsbuildPlugins,
  RsbuildPluginAPI,
} from './types';

export type {
  RsbuildMode,
  RsbuildEntry,
  RsbuildTarget,
  RsbuildContext,
  RsbuildInstance,
  CreateRsbuildOptions,
  InspectConfigOptions,
  // Subtypes of Config
  Minify,
  Polyfill,
  PrintUrls,
  PublicDir,
  Decorators,
  RspackRule,
  WatchFiles,
  CSSModules,
  CrossOrigin,
  ConsoleType,
  SplitChunks,
  BundlerChain,
  ClientConfig,
  ScriptInject,
  ChainedConfig,
  PostCSSPlugin,
  ScriptLoading,
  LegalComments,
  AliasStrategy,
  FilenameConfig,
  DistPathConfig,
  OutputStructure,
  ChainIdentifier,
  PreconnectOption,
  CSSLoaderOptions,
  ModifyChainUtils,
  StyleLoaderOptions,
  PostCSSLoaderOptions,
  ChainedConfigWithUtils,
  ModifyRspackConfigUtils,
  CSSModulesLocalsConvention,
  // Hook Callback Types
  OnExitFn,
  OnAfterBuildFn,
  OnAfterCreateCompilerFn,
  OnAfterStartDevServerFn,
  OnAfterStartProdServerFn,
  OnBeforeBuildFn,
  OnBeforeStartDevServerFn,
  OnBeforeStartProdServerFn,
  OnBeforeCreateCompilerFn,
  OnCloseDevServerFn,
  OnDevCompileDoneFn,
  ModifyBundlerChainFn,
  ModifyRspackConfigFn,
  ModifyRsbuildConfigFn,
  TransformFn,
  TransformHandler,
} from '@rsbuild/shared';

export {
  /**
   * @private
   */
  __internalHelper,
};
