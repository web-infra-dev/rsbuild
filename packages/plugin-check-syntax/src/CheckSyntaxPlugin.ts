import { resolve } from 'node:path';
import type { Rspack } from '@rsbuild/core';
import { JS_REGEX, browserslistToESVersion, fse } from '@rsbuild/shared';
import { parse } from 'acorn';
import {
  checkIsExcludeSource,
  generateError,
  generateHtmlScripts,
  printErrors,
} from './helpers';
import type {
  AcornParseError,
  CheckSyntaxExclude,
  CheckSyntaxOptions,
  ECMASyntaxError,
  EcmaVersion,
} from './types';

type Compiler = Rspack.Compiler;
type Compilation = Rspack.Compilation;

const HTML_REGEX = /\.html$/;

export class CheckSyntaxPlugin {
  errors: ECMASyntaxError[] = [];

  ecmaVersion: EcmaVersion;

  targets: string[];

  rootPath: string;

  exclude: CheckSyntaxExclude | undefined;

  constructor(
    options: CheckSyntaxOptions &
      Required<Pick<CheckSyntaxOptions, 'targets'>> & {
        rootPath: string;
      },
  ) {
    this.targets = options.targets;
    this.exclude = options.exclude;
    this.rootPath = options.rootPath;
    this.ecmaVersion =
      options.ecmaVersion || browserslistToESVersion(this.targets);
  }

  apply(complier: Compiler) {
    complier.hooks.afterEmit.tapPromise(
      CheckSyntaxPlugin.name,
      async (compilation: Compilation) => {
        const outputPath = compilation.outputOptions.path || 'dist';

        // not support compilation.emittedAssets in Rspack
        const emittedAssets = compilation
          .getAssets()
          .filter((a) => a.source)
          .map((a) => {
            // remove query from name
            const resourcePath = a.name.split('?')[0];
            return resolve(outputPath, resourcePath);
          });

        const files = emittedAssets.filter(
          (assets) => HTML_REGEX.test(assets) || JS_REGEX.test(assets),
        );
        await Promise.all(
          files.map(async (file) => {
            await this.check(file);
          }),
        );

        printErrors(this.errors);
      },
    );
  }

  private async check(filepath: string) {
    if (HTML_REGEX.test(filepath)) {
      const htmlScripts = await generateHtmlScripts(filepath);
      await Promise.all(
        htmlScripts.map(async (script) => {
          if (!checkIsExcludeSource(filepath, this.exclude)) {
            await this.tryParse(filepath, script);
          }
        }),
      );
    }

    if (JS_REGEX.test(filepath)) {
      const jsScript = await fse.readFile(filepath, 'utf-8');
      await this.tryParse(filepath, jsScript);
    }
  }

  private async tryParse(filepath: string, code: string) {
    try {
      parse(code, { ecmaVersion: this.ecmaVersion });
    } catch (_: unknown) {
      const err = _ as AcornParseError;

      const error = await generateError({
        err,
        code,
        filepath,
        exclude: this.exclude,
        rootPath: this.rootPath,
      });

      if (error) {
        this.errors.push(error);
      }
    }
  }
}
