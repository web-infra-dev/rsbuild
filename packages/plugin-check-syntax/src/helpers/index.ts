export * from './generateError';
export * from './printErrors';
export * from './generateHtmlScripts';
export * from './utils';
