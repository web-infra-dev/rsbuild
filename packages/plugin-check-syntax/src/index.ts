import type { RsbuildPlugin } from '@rsbuild/core';
import {
  type NormalizedConfig,
  type RsbuildTarget,
  getBrowserslistWithDefault,
} from '@rsbuild/shared';
import type { CheckSyntaxOptions } from './types';

export type PluginCheckSyntaxOptions = CheckSyntaxOptions;

async function getTargets(
  rootPath: string,
  rsbuildConfig: NormalizedConfig,
  rsbuildTarget: RsbuildTarget,
  checkSyntax: CheckSyntaxOptions,
) {
  return (
    checkSyntax.targets ??
    (await getBrowserslistWithDefault(rootPath, rsbuildConfig, rsbuildTarget))
  );
}

export function pluginCheckSyntax(
  options: PluginCheckSyntaxOptions = {},
): RsbuildPlugin {
  return {
    name: 'rsbuild:check-syntax',

    setup(api) {
      api.modifyBundlerChain(async (chain, { isDev, target }) => {
        if (isDev || target !== 'web') {
          return;
        }

        const { rootPath } = api.context;
        const rsbuildConfig = api.getNormalizedConfig();

        const targets = await getTargets(
          rootPath,
          rsbuildConfig,
          target,
          options,
        );
        const { CheckSyntaxPlugin } = await import('./CheckSyntaxPlugin');

        chain.plugin(CheckSyntaxPlugin.name).use(CheckSyntaxPlugin, [
          {
            targets,
            rootPath,
            ...(typeof options === 'object' ? options : {}),
          },
        ]);
      });
    },
  };
}
