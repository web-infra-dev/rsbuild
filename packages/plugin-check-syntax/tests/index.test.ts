import { createStubRsbuild } from '@scripts/test-helper';
import { pluginCheckSyntax } from '../src/index';

beforeAll(() => {
  const { NODE_ENV } = process.env;
  process.env.NODE_ENV = 'production';

  return () => {
    process.env.NODE_ENV = NODE_ENV;
  };
});

describe('plugin-check-syntax', () => {
  it('should add check-syntax plugin properly', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginCheckSyntax()],
    });

    const config = await rsbuild.unwrapConfig();

    expect(config).toMatchSnapshot();
  });

  it('should not add check-syntax plugin when target node', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginCheckSyntax()],
      rsbuildConfig: {
        output: {
          target: 'node',
        },
      },
    });

    const config = await rsbuild.unwrapConfig();

    expect(config).toMatchSnapshot();
  });

  it('should use default browserslist as targets when only set checkSyntax.exclude', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [
        pluginCheckSyntax({
          exclude: [/$.html/],
        }),
      ],
      rsbuildConfig: {
        output: {
          overrideBrowserslist: [
            'iOS 9',
            'Android 4.4',
            'last 2 versions',
            '> 0.2%',
            'not dead',
          ],
        },
      },
    });

    const config = await rsbuild.unwrapConfig();

    expect(config).toMatchSnapshot();
  });
});
