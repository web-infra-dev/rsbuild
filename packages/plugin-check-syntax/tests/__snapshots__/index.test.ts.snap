// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugin-check-syntax > should add check-syntax plugin properly 1`] = `
{
  "plugins": [
    CheckSyntaxPlugin {
      "ecmaVersion": 2017,
      "errors": [],
      "exclude": undefined,
      "rootPath": "<ROOT>/packages/plugin-check-syntax/tests",
      "targets": [
        "chrome >= 87",
        "edge >= 88",
        "firefox >= 78",
        "safari >= 14",
      ],
    },
  ],
}
`;

exports[`plugin-check-syntax > should not add check-syntax plugin when target node 1`] = `{}`;

exports[`plugin-check-syntax > should use default browserslist as targets when only set checkSyntax.exclude 1`] = `
{
  "plugins": [
    CheckSyntaxPlugin {
      "ecmaVersion": 5,
      "errors": [],
      "exclude": [
        /\\$\\.html/,
      ],
      "rootPath": "<ROOT>/packages/plugin-check-syntax/tests",
      "targets": [
        "iOS 9",
        "Android 4.4",
        "last 2 versions",
        "> 0.2%",
        "not dead",
      ],
    },
  ],
}
`;
